feat(parser): add support for Lambda Function URL

ran-isenberg opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number: #1425**

## Summary

### Changes
added envelope and model for lambda function url as described at https://github.com/awslabs/aws-lambda-powertools-python/issues/1425

fixed a bug in api gwv2 model where cognitoIdentity wasnt optional.

I was unable to fix the docs for the parser due to a linter bug, will open an issue.

### User experience

Can now use an envelope for lambda function url or extend the model.


<details>
<summary>Is this a breaking change? No.</summary>

**RFC issue number**:
#1425
Checklist:

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

ran-isenberg committed:
feat: Add Parser official support for Lambda Function UR

ran-isenberg commented:
@heitorlessa  i wasnt able to push the docs update.
They pass the docs-linter stage but fail on the pre-commit on this error:
docs/utilities/parser.md:586 MD043/required-headings/required-headers Required heading structure [Context: "## Getting started"]

leandrodamascena commented:
Hi @ran-isenberg.. I'm refactoring all examples and for now you can add this comment after the description line..

`<!-- markdownlint-disable MD043 -->
`

like this: https://github.com/awslabs/aws-lambda-powertools-python/commit/7be425e05f2c193a20c1558de867dc39b81cabb3#diff-aa083264efcaf3125cc0560e9b88cce2c69e4becd8ea39211057634ec3f56bd9R6

heitorlessa commented:
Thanks a lot @ran-isenberg !! As @leandrodamascena shared, this is the new doc linting to enforce a consistent style, you can ignore for now with that string at the top of the file since we're refactoring all of it (structure, snippets to be way more realistic, run as-is). 

@leandrodamascena could you please fix that mypy issue by installing `snappy` as a dev dependency and ignoring the lack of types for snappy? We missed in the last PR as I accidentally didn't allow GH Actions to run (part of the new repo hardening)

ran-isenberg commented:
> Hi @ran-isenberg.. I'm refactoring all examples and for now you can add this comment after the description line..
> 
> `<!-- markdownlint-disable MD043 --> `
> 
> like this: [7be425e#diff-aa083264efcaf3125cc0560e9b88cce2c69e4becd8ea39211057634ec3f56bd9R6](https://github.com/awslabs/aws-lambda-powertools-python/commit/7be425e05f2c193a20c1558de867dc39b81cabb3#diff-aa083264efcaf3125cc0560e9b88cce2c69e4becd8ea39211057634ec3f56bd9R6)

Tried it but it doesnt work for me, cant commit. I added this linter to my own project and just removed this annoying check from the config :)

ran-isenberg committed:
Update parser.md

ran-isenberg commented:
pushed the docs ;)

ran-isenberg commented:
@heitorlessa  it failed on the snappy stuff. I updated from develop branch, hopefully it will work now

heitorlessa commented:
Yes it was fixed on #1446, my other PR was impacted and working after the
merge

On Fri, 12 Aug 2022 at 08:20, Ran Isenberg ***@***.***> wrote:

> @heitorlessa <https://github.com/heitorlessa> it failed on the snappy
> stuff. I updated from develop branch, hopefully it will work now
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/awslabs/aws-lambda-powertools-python/pull/1442#issuecomment-1212768337>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZPQBGY3STSYALQBFFT4I3VYXUKFANCNFSM56IX7WWQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

heitorlessa commented:
@rubenfonseca assigned it to you as you fixed the quirks of Function URL.

@ran-isenberg CI passed. Ruben is back on Monday from holidays and should get this reviewed and merged next week.

Thank you again

rubenfonseca commented:
Just rebased with master before the review

rubenfonseca commented on aws_lambda_powertools/utilities/parser/envelopes/lambda_function_url.py:
```diff
@@ -0,0 +1,32 @@
+        Any
+            Parsed detail payload with model provided
+        """
+        logger.debug(f"Parsing incoming data with Lambda function URL {LambdaFunctionUrlModel}")
```
```suggestion
        logger.debug(f"Parsing incoming data with Lambda function URL model {LambdaFunctionUrlModel}")
```

rubenfonseca commented on docs/utilities/parser.md:
```diff
@@ -166,6 +167,7 @@ Parser comes with the following built-in models:
 | **SnsModel**                    | Lambda Event Source payload for Amazon Simple Notification Service |
 | **APIGatewayProxyEventModel**   | Lambda Event Source payload for Amazon API Gateway                 |
 | **APIGatewayProxyEventV2Model** | Lambda Event Source payload for Amazon API Gateway v2 payload      |
+| **LambdaFunctionUrlModel**      | Lambda Event Source payload for Lambda function URL payload        |
```
```suggestion
| **LambdaFunctionUrlModel**      | Lambda Event Source payload for Lambda Function URL payload        |
```

rubenfonseca reviewed (changes requested):
Hi @ran-isenberg thank you for this! Just left two small suggestions. Let me know if it makes sense so we can merge.

ran-isenberg committed:
Update aws_lambda_powertools/utilities/parser/envelopes/lambda_function_url.py

Co-authored-by: Ruben Fonseca <fonseka@gmail.com>

ran-isenberg committed:
Update docs/utilities/parser.md

Co-authored-by: Ruben Fonseca <fonseka@gmail.com>

leandrodamascena commented:
> Hi @ran-isenberg thank you for this! Just left two small suggestions. Let me know if it makes sense so we can merge.

Hi!! Just in case of better understanding of the project I did a double check and everything seems fine to merge. I ran these changes locally and it went fine.

ran-isenberg committed:
Merge branch 'develop' into func_url

ran-isenberg commented:
@heitorlessa  waiting for the merge :)

heitorlessa commented:
@rubenfonseca will be merging ;) Thanks a lot everyone!!

rubenfonseca commented:
Thanks again Ran! Can't wait to get this out soon :)
